### PR TITLE
Cosmos Stargate: protobuf serialization

### LINKED ISSUE #1446: [Cosmos] Serialized tx for new stargate api endpoint
From https://v1.cosmos.network/rpc/v0.41.4, `/cosmos/tx/v1beta1/txs` endpoint

### PR DESCRIPTION
**Pull request by catenocrypt**
## Description

Cosmos new Stargate serialization and signing.
Old way:  Message is constructed in json, signed, and broadcast as json using `/txs` RPC.
New way:  Message is constructed in binary protobuf-serialization form, signed, and broadcast using `/cosmos/tx/v1beta1/txs`.  Supported since SDK version v0.40.

Adaptation guide:

1. Add SigningMode = Protobuf to Signing input:

```swift
    $0.signingMode = .protobuf;
```
```kotlin
    signingMode = SigningMode.Protobuf
```

2. Change API endpoint to `/cosmos/tx/v1beta1/txs`, using field `SigningOutput.serializedBase64`

Fixes #1446

Support is added in a compatible way, new `SigningInput.signing_mode` field has default value `JSON` for old, `Protobuf` for new mode.  Old Json signing can be deprecated once all Cosmos-based blockchains adapt to the new Stargate API.

Proto files are taken from the Cosmos SDK, and they are kept separate from the 'usual' proto files used for swift/kotlin interfacing.  The internal proto files are used for serialization for RPC messages.

Cryptoorg tests are adapted to use the new mode.

## Testing instructions

Internal protobuf tests (`CosmosProtobuf`), incl. with test vectors from Cosmos doc.
Real-world Cryptoorg transfer test (`TWAnySignerCryptoorg.SignTx_Proto_BCB213`).
Unit tests, `*osmos*:*rypto*`.

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain
- [x] If there is a related Issue, mention it in the description (e.g. Fixes #<issue_number> ).

### CODE REVIEW

**Comment by catenocrypt** [comment]
Tested with real-world Atom transfer TX.
To do: check coverage.

**Comment by catenocrypt** [comment]
Coverage OK, ready for merge.

File: tests/Cosmos/Protobuf/Article.proto [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  NEWS = 2;
+};
+
+enum Review {
+  REVIEW_UNSPECIFIED = 0;
+  ACCEPTED = 1;
+  REJECTED = 2;
+};
+
+message Article {
```
Comment (hewigovens) [inline tests/Cosmos/Protobuf/Article.proto:16]: maybe we don't need this? we already have `MsgSend` test

File: tools/generate-files [hunk 2]
```diff
@@ -56,9 +56,11 @@ else
     "$PROTOC" -I=$PREFIX/include -I=src/proto --cpp_out=src/proto --java_out=jni/java src/proto/*.proto
 fi
 
-# Generate internal message protocol Protobuf files -- not every time
+# Generate internal message protocol Protobuf files
 "$PROTOC" -I=$PREFIX/include -I=src/Tron/Protobuf --cpp_out=src/Tron/Protobuf src/Tron/Protobuf/*.proto
 "$PROTOC" -I=$PREFIX/include -I=src/Zilliqa/Protobuf --cpp_out=src/Zilliqa/Protobuf src/Zilliqa/Protobuf/*.proto
+"$PROTOC" -I=$PREFIX/include -I=src/Cosmos/Protobuf --cpp_out=src/Cosmos/Protobuf src/Cosmos/Protobuf/*.proto
+"$PROTOC" -I=$PREFIX/include -I=tests/Cosmos/Protobuf --cpp_out=tests/Cosmos/Protobuf tests/Cosmos/Protobuf/*.proto
```
Comment (hewigovens) [inline tools/generate-files:63]: let's remove test purpose photos

File: src/proto/Cosmos.proto [hunk 3]
```diff
@@ -75,25 +75,43 @@ message Message {
...
 message SigningOutput {
     // Signature
     bytes signature = 1;
-    // Signed transaction in JSON.
+
+    // Signed transaction in JSON (pre-Stargate case)
     string json = 2;
+
+    // Signed transaction in protobuf form (Stargate case)
+    bytes serialized = 3;
```
Comment (hewigovens) [inline src/proto/Cosmos.proto:110]: I think we can only keep one base64 encoded `serialized` that api endpoint expects?

**Review by hewigovens — APPROVED** [review]
Looks good, left some comments

File: tests/Cosmos/Protobuf/Article.proto (hunk 1 above)
Comment (catenocrypt) [reply tests/Cosmos/Protobuf/Article.proto:16]: I did the test from the Cosmos SDK ADR; this has some test for some special cases.  It's true that rest of tests indirectly test serialization, but not necessarily all aspects.  It's important to have serialization identical to each byte.  I'd keep it.

File: tools/generate-files (hunk 2 above)
Comment (catenocrypt) [reply tools/generate-files:63]: See above.

File: src/proto/Cosmos.proto (hunk 3 above)
Comment (catenocrypt) [reply src/proto/Cosmos.proto:110]: The raw data version is only good for debugging, I remove it.

**Review by vikmeup — APPROVED** [review]

**Pull request merged**